File: Editorial/104_Maximum_Depth_of_Binary_Tree/C/3_Iteration.c
struct Pair {
    struct TreeNode* node;
    int depth;
};
struct Pair* pair_create(struct TreeNode* node, int depth) {
    struct Pair* pair = malloc(sizeof(struct Pair));
    pair->node = node;
    pair->depth = depth;
    return pair;
}
int maxDepth(struct TreeNode* root) {
    if (root == NULL) {
        return 0;
    }
    struct Pair** stack = malloc(10000 * sizeof(struct TreeNode*));
    int top = 0;
    stack[top++] = pair_create(root, 1);
    int max_depth = 0;
    while (top != 0) {
        struct Pair* pair = stack[--top];
        int c_depth = pair->depth;
        struct TreeNode* c_node = pair->node;
        max_depth = c_depth > max_depth ? c_depth : max_depth;
        free(pair);
        if (c_node->left != NULL) {
            stack[top++] = pair_create(c_node->left, c_depth + 1);
        }
        if (c_node->right != NULL) {
            stack[top++] = pair_create(c_node->right, c_depth + 1);
        }
    }
    free(stack);
    return max_depth;
}

File: Editorial/104_Maximum_Depth_of_Binary_Tree/Go/3_Iteration.go
func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    stack := []struct {
        depth int
        node  *TreeNode
    }{{1, root}}
    depth := 0
    for len(stack) != 0 {
        current := stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        if current.node != nil {
            if depth < current.depth {
                depth = current.depth
            }
            stack = append(stack, struct {
                depth int
                node  *TreeNode
            }{current.depth + 1, current.node.Left})
            stack = append(stack, struct {
                depth int
                node  *TreeNode
            }{current.depth + 1, current.node.Right})
        }
    }
    return depth
}
